### DISCUSSION TITLE
Is this allocation free (gc free)?

### DISCUSSION THREAD

**Discussion by mason124124** [simdjson/simdjson-java#58] [Q&A] [answered]
Thanks for sharing this. I can see the benchmark is pretty amazing, however, may I ask is this allocation free?

**Comment by piotrrzysko** [comment, marked as answer]
Currently, it is not GC-free, because, for example, it creates [JsonValue](https://github.com/simdjson/simdjson-java/blob/main/src/main/java/org/simdjson/JsonValue.java) objects for every parsed JSON. However, the main data structures are allocated only once (when a `SimdJsonParser` instance is created), so I belive it wouldn't be difficult to introduce a GC-free version.

**Reply by mason124124** [reply]
thanks for the reply. I think it would be extremely helpful to introduce a gc free version, and it will improve the popularity for this project a lot. Currently, all the crypto trading industry are dying for a gc free version of json parser for crypto hft. Obviously, we all implement our own verison of gc free parser. but would be much better to have support here. Also, if it would support using java bytebuffer and agrona direct buffer, to allow manipulation of offheap would be ideal as well

**Reply by prdoyle** [reply]
May I ask what is the motivation for a GC free one? Is it to prevent GC pauses?

If it's straight up performance, then it's possible making `JsonValue` into a record might help a little. Record fields can be optimized by the JIT in ways that regular final instance fields cannot, and down the line, hopefully it could be a "value record" which then wouldn't need allocation at all.

**Reply by piotrrzysko** [reply]
There are applications where GC pauses are unacceptable. One example that comes to mind is high-frequency trading.

**Reply by mason124124** [reply]
yes, correct, this is for trading
